packages/ember/tests/routing/decoupled_basic_test.js

```diff
@@ -326,4 +327,217 @@ class extends ApplicationTestCase {
+        set(this.controllerFor('home'), 'hours', emberA([
+          'Monday through Friday: 9am to 5pm',
+          'Saturday: Noon to Midnight',
+          'Sunday: Noon to 6pm'
+        ]));
+      }
+    }));
+
+    this.addTemplate('home', '<ul>{{#each hours as |entry|}}<li>{{entry}}</li>{{/each}}</ul>');
+
+    return this.visit('/').then(() => {
+      let text = this.$('ul li:nth-child(3)').text();
+
+      assert.equal(text, 'Sunday: Noon to 6pm',
+        'The template was rendered with the hours context');
+    });
+  }
+
+  [`@test The Homepage with a computed context that does not get overridden`](assert) {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+    });
+
+    this.add('controller:home', Controller.extend({
+      model: computed(function() {
+        return emberA([
+          'Monday through Friday: 9am to 5pm',
+          'Saturday: Noon to Midnight',
+          'Sunday: Noon to 6pm'
+        ]);
+      })
+    }));
+
+    this.addTemplate('home', '<ul>{{#each model as |passage|}}<li>{{passage}}</li>{{/each}}</ul>');
+
+    return this.visit('/').then(() => {
+      let text = this.$('ul li:nth-child(3)').text();
+
+      assert.equal(text, 'Sunday: Noon to 6pm',
+        'The template was rendered with the context intact');
+    });
+  }
+
+  [`@test The Homepage getting its controller context via model`](assert) {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+    });
+
+    this.add('route:home', Route.extend({
+      model() {
+        return emberA([
+          'Monday through Friday: 9am to 5pm',
+          'Saturday: Noon to Midnight',
+          'Sunday: Noon to 6pm'
+        ]);
+      },
+
+      setupController(controller, model) {
+        assert.equal(this.controllerFor('home'), controller);
+
+        set(this.controllerFor('home'), 'hours', model);
+      }
+    }));
+
+    this.addTemplate('home', '<ul>{{#each hours as |entry|}}<li>{{entry}}</li>{{/each}}</ul>');
+
+    return this.visit('/').then(() => {
+      let text = this.$('ul li:nth-child(3)').text();
+
+      assert.equal(text, 'Sunday: Noon to 6pm',
+        'The template was rendered with the hours context');
+    });
+
+  }
+
+  [`The Specials Page getting its controller context by deserializing the params hash`](assert) {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+      this.route('special', { path: '/specials/:menu_item_id' });
+    });
+
+    this.add('route:special',  Route.extend({
+      model(params) {
+        return EmberObject.create({
+          menuItemId: params.menu_item_id
+        });
+      },
+
+      setupController(controller, model) {
+        set(controller, 'model', model);
+      }
+    }));
+
+    this.addTemplate('special', '<p>{{model.menuItemId}}</p>');
+
+    return this.visit('/').then(() => {
+      this.add('controller:special', Controller.extend());
+
+      return this.visit('/specials/1').then(() => {
```

Comment:
Chain promises, don't nest. For example:

```js
return this.visit('/').then(() => {
  this.add('controller:special', Controller.extend());
  return this.visit('/specials/1');
}).then(() => {
  let test = this.$('p').text();
  assert.equal(text, '1',
    'the model was used to render the template'
  );
});
```

Reply:
Actually I think you can just remove `this.add('controller:special', Controller.extend());` here.

I don't see the relevance here of visiting `/` before the route that is being tested either. IMO you can drop the route `this.route('home', { path: '/' });` above then just make this:

```js
return this.visit('/specials/1').then(() => {
  let test = this.$('p').text();
  assert.equal(text, '1',
    'the model was used to render the template'
  );
});
```